File: robin_scalper/exchange/paper_broker.py
"""
模拟盘实现：内存里维护持仓和成交，吃最新行情价成交。
"""
from __future__ import annotations
import time
import uuid
import asyncio
from typing import Dict, List, Optional

from .base import BrokerBase, Order, Position
from ..core.logger import LogBuffer
from ..core.bus import EventBus


class PaperBroker(BrokerBase):
    def __init__(self, symbol: str, bus: EventBus, log: LogBuffer,
                 leverage: int = 10, margin_mode: str = "ISOLATED"):
        super().__init__(symbol, leverage, margin_mode)
        self.bus = bus
        self.log = log
        # 按 side 持仓
        self.positions: Dict[str, Position] = {}
        self.last_px: float = 0.0
        # 订阅 tick，更新价格
        bus.subscribe("tick", self._on_tick)

    def mode(self) -> str:
        return "PAPER"

    def _on_tick(self, env: dict) -> None:
        self.last_px = float(env["data"]["price"])
        for p in self.positions.values():
            if p.side == "LONG":
                p.unrealized_pnl = (self.last_px - p.entry_price) * p.qty
            else:
                p.unrealized_pnl = (p.entry_price - self.last_px) * p.qty

    async def set_leverage(self) -> None:
        return  # 模拟盘不需要

    async def get_mark_price(self) -> float:
        return self.last_px

    def _position_side(self, side: str) -> str:
        return "LONG" if side == "BUY" else "SHORT"

    async def open_market(self, side: str, qty: float) -> Order:
        # 模拟盘有 last_px=0 的可能，让策略重试
        if self.last_px <= 0:
            self.log.warn("模拟盘尚未收到行情价")
            return Order(id="", symbol=self.symbol, side=side,
                         position_side=self._position_side(side),
                         type="MARKET", qty=qty, price=0, status="REJECTED")
        pos_side = self._position_side(side)
        # 同一方向累加
        if pos_side in self.positions:
            p = self.positions[pos_side]
            new_qty = p.qty + qty
            p.entry_price = (p.entry_price * p.qty + self.last_px * qty) / new_qty
            p.qty = new_qty
        else:
            self.positions[pos_side] = Position(
                symbol=self.symbol, side=pos_side, qty=qty,
                entry_price=self.last_px, leverage=self.leverage,
                margin_type=self.margin_mode,
                client_id=f"sim-{uuid.uuid4().hex[:8]}",
            )
        order = Order(
            id=uuid.uuid4().hex,
            symbol=self.symbol, side=side, position_side=pos_side,
            type="MARKET", qty=qty, price=self.last_px, status="FILLED",
            client_id=self.positions[pos_side].client_id,
        )
        self.log.trade(f"[PAPER] OPEN {side} qty={qty} px={self.last_px} id={order.id}")
        return order

    async def close_market(self, side: str, qty: Optional[float] = None) -> List[Order]:
        pos_side = self._position_side(side)
        p = self.positions.get(pos_side)
        if p is None or p.qty <= 0:
            return []
        close_qty = p.qty if qty is None else min(qty, p.qty)
        price = self.last_px
        order = Order(
            id=uuid.uuid4().hex, symbol=self.symbol, side=side,
            position_side=pos_side, type="MARKET", qty=close_qty,
            price=price, status="FILLED", reduce_only=True,
        )
        p.qty -= close_qty
        if p.qty <= 1e-9:
            del self.positions[pos_side]
        self.log.trade(f"[PAPER] CLOSE {side} qty={close_qty} px={price} id={order.id}")
        return [order]

    async def get_positions(self) -> List[Position]:
        return [p for p in self.positions.values() if p.qty > 0]

File: robin-client/src/core/paper-broker.ts
// Paper Broker - 模拟盘 - 对应 Python 版本的 paper_broker.py

import { BrokerBase } from './broker-base';
import { Order, Position } from './types';

export class PaperBroker extends BrokerBase {
  private positions: Map<string, Position> = new Map();
  private lastPx = 0;
  private onTick?: (price: number) => void;

  constructor(
    symbol: string,
    leverage: number = 10,
    marginMode: 'ISOLATED' | 'CROSSED' = 'ISOLATED',
    onTick?: (price: number) => void
  ) {
    super(symbol, leverage, marginMode);
    this.onTick = onTick;
  }

  mode(): 'PAPER' | 'LIVE' {
    return 'PAPER';
  }

  updatePrice(price: number): void {
    this.lastPx = price;
    // 更新未实现盈亏
    for (const p of this.positions.values()) {
      if (p.side === 'LONG') {
        p.unrealized_pnl = (price - p.entry_price) * p.qty;
      } else {
        p.unrealized_pnl = (p.entry_price - price) * p.qty;
      }
    }
    this.onTick?.(price);
  }

  async setLeverage(): Promise<void> {
    // 模拟盘不需要
  }

  async getMarkPrice(): Promise<number> {
    return this.lastPx;
  }

  private posSide(side: 'BUY' | 'SELL'): 'LONG' | 'SHORT' {
    return side === 'BUY' ? 'LONG' : 'SHORT';
  }

  async openMarket(side: 'BUY' | 'SELL', qty: number): Promise<Order> {
    if (this.lastPx <= 0) {
      return {
        id: '',
        symbol: this.symbol,
        side,
        position_side: this.posSide(side),
        type: 'MARKET',
        qty,
        price: 0,
        reduce_only: false,
        client_id: '',
        ts: Date.now(),
        status: 'REJECTED',
      };
    }

    const posSide = this.posSide(side);
    const price = this.lastPx;

    if (this.positions.has(posSide)) {
      const p = this.positions.get(posSide)!;
      const newQty = p.qty + qty;
      p.entry_price = (p.entry_price * p.qty + price * qty) / newQty;
      p.qty = newQty;
    } else {
      const id = crypto.randomUUID();
      this.positions.set(posSide, {
        symbol: this.symbol,
        side: posSide,
        qty,
        entry_price: price,
        unrealized_pnl: 0,
        leverage: this.leverage,
        margin_type: this.marginMode,
        client_id: `sim-${id.slice(0, 8)}`,
        opened_at: Date.now(),
      });
    }

    return {
      id: crypto.randomUUID(),
      symbol: this.symbol,
      side,
      position_side: posSide,
      type: 'MARKET',
      qty,
      price,
      reduce_only: false,
      client_id: this.positions.get(posSide)!.client_id,
      ts: Date.now(),
      status: 'FILLED',
    };
  }

  async closeMarket(side: 'BUY' | 'SELL', qty?: number): Promise<Order[]> {
    const posSide = this.posSide(side);
    const p = this.positions.get(posSide);
    if (!p || p.qty <= 0) return [];

    const closeQty = qty === undefined ? p.qty : Math.min(qty, p.qty);
    const price = this.lastPx;

    p.qty -= closeQty;
    if (p.qty <= 1e-9) {
      this.positions.delete(posSide);
    }

    return [{
      id: crypto.randomUUID(),
      symbol: this.symbol,
      side,
      position_side: posSide,
      type: 'MARKET',
      qty: closeQty,
      price,
      reduce_only: true,
      client_id: '',
      ts: Date.now(),
      status: 'FILLED',
    }];
  }

  async getPositions(): Promise<Position[]> {
    return [...this.positions.values()].filter(p => p.qty > 0);
  }
}
